physical/raft/snapshot.go

```diff
@@ -151,69 +156,154 @@ func (f *BoltSnapshotStore) getBoltSnapshotMeta() (*raft.SnapshotMeta, error) {
+		return f.openFromFSM()
+	}
 
-		// Stream data out of the FSM to calculate the size
-		var writeCloser *io.PipeWriter
-		readCloser, writeCloser = io.Pipe()
-		metaReadCloser, metaWriteCloser := io.Pipe()
-		go func() {
-			f.fsm.writeTo(context.Background(), metaWriteCloser, writeCloser)
-		}()
-
-		// Compute the size
-		n, err := io.Copy(ioutil.Discard, metaReadCloser)
-		if err != nil {
-			f.logger.Error("failed to read state file", "error", err)
-			metaReadCloser.Close()
-			readCloser.Close()
-			return nil, nil, err
-		}
+	return f.openFromFile(id)
+}
 
-		meta.Size = n
+func (f *BoltSnapshotStore) openFromFSM() (*raft.SnapshotMeta, io.ReadCloser, error) {
+	meta, err := f.getMetaFromFSM()
+	if err != nil {
+		return nil, nil, err
+	}
+	// If we don't have any data return an error
+	if meta.Index == 0 {
+		return nil, nil, errors.New("no snapshot data")
+	}
 
-	default:
-		var err error
-		meta, readCloser, err = f.fileSnapStore.Open(id)
-		if err != nil {
-			return nil, nil, err
+	// Stream data out of the FSM to calculate the size
+	readCloser, writeCloser := io.Pipe()
+	metaReadCloser, metaWriteCloser := io.Pipe()
+	go func() {
+		f.fsm.writeTo(context.Background(), metaWriteCloser, writeCloser)
+	}()
+
+	// Compute the size
+	n, err := io.Copy(ioutil.Discard, metaReadCloser)
+	if err != nil {
+		f.logger.Error("failed to read state file", "error", err)
+		metaReadCloser.Close()
+		readCloser.Close()
+		return nil, nil, err
+	}
+
+	meta.Size = n
+
+	return meta, readCloser, nil
+}
+
+func (f *BoltSnapshotStore) getMetaFromDB(id string) (*raft.SnapshotMeta, error) {
+	if len(id) == 0 {
+		return nil, errors.New("can not open empty snapshot ID")
+	}
+
+	filename := filepath.Join(f.path, id, databaseFilename)
+	boltDB, err := bolt.Open(filename, 0666, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return nil, err
+	}
+	defer boltDB.Close()
+
+	meta := &raft.SnapshotMeta{
+		Version: 1,
+		ID:      id,
+	}
+
+	err = boltDB.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(configBucketName)
+		val := b.Get(latestIndexKey)
+		if val != nil {
+			var snapshotIndexes IndexValue
+			err := proto.Unmarshal(val, &snapshotIndexes)
+			if err != nil {
+				return err
+			}
+
+			meta.Index = snapshotIndexes.Index
+			meta.Term = snapshotIndexes.Term
 		}
 
-		readCloser = &boltSnapshotMetadataReader{
-			meta:       meta,
-			ReadCloser: readCloser,
+		// Read in our latest config and populate it inmemory
+		val = b.Get(latestConfigKey)
+		if val != nil {
+			var config ConfigurationValue
+			err := proto.Unmarshal(val, &config)
+			if err != nil {
+				return err
+			}
+
+			meta.ConfigurationIndex, meta.Configuration = protoConfigurationToRaftConfiguration(&config)
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return meta, nil
+}
+
+func (f *BoltSnapshotStore) openFromFile(id string) (*raft.SnapshotMeta, io.ReadCloser, error) {
+	// TODO: should we insted have a separate metadata file so we can checksum
```

Comment:
Should this TODO be addressed/dismissed?

Reply:
I went ahead and removed this. We can always add an extra file later, the bolt db does have its own checksums.